### PR TITLE
Added GitHub actions to validate issue number in PR bodies

### LINKED ISSUE #757: Github Actions: Auto-Close Issues if the PR is merged
# Description

In many of the cases, I found that the Issues are not closed when the PR is merged. We can have a Github Action which can take care of closing the Issues when the PR is merged automatically. Few edge cases we can cover are: 

- Does the PR contain an Issue? If not, remind the contributor to link the Issue to the PR (You can add labels as well)
- Is the Issue assigned to more than one person? If yes, then we don't need to close the Issue pre-maturely.

| name            | problem/motivation                                                  | possible solution  | assignees                                                      |
| --------------- | ------------------------------------------------------ | ------- | -------------------------------------------------------------- |
| Auto-Close Bot | Close pending Issues automatically once the PR has been merged | Github Actions Marketplace have some pretty good actions for the same :) | |

### PR DESCRIPTION
**Pull request by XZANATOL**
# Description

There was a need to automate PR body validation, check for valid issue numbers, and apply a corresponding decission if true or false. (This is the first action out of two)

The GitHub action can be viewed in action from [here](https://github.com/XZANATOL/Github-actions-test/pull/18)

Fixes #757 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation Update

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have created a helpful and easy to understand `README.md`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by XZANATOL** [comment]
`` PR has a valid Issue? / checker (pull_request) `` has failed because the action doesn't have access to the secret envs yet. (will have after merging)

File: .github/actions/PR_has_a_valid_issue/main.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import os, re, requests
+
+pattern = "#\d+"  # Used pattern to search for Issues.
```
Comment (vybhav72954) [inline .github/actions/PR_has_a_valid_issue/main.py:3]: In this Repo, we follow this style:

```
# Used pattern to search for Issues.
pattern = "#\d+"
```
Can you change accordingly?

File: .github/workflows/Check if issue exists.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        with:
+          prbody: ${{ github.event.pull_request.body }}
+          prurl: ${{ github.event.pull_request.url }}
+
+      - name: "Validation action on 1"
+        if: ${{ steps.validator.outputs.valid == 1 }}
+        run: |
+          echo $Github_Token > mytoken.txt
+          gh auth login --with-token < mytoken.txt
+          gh pr edit $PRNUM --add-label "PR:Ready-to-Merge"
```
Comment (vybhav72954) [inline .github/workflows/Check if issue exists.yml:25]: Change this label to `PR:Ready-to-Review`

File: .github/actions/PR_has_a_valid_issue/main.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        else:
+            print("Couldn't find Open flag")
+else:
+    print("Invalid Response Code obtained - error code: {}".format(response.status_code))
+
+print("Valid flag is:", valid_code)
+
+print(f"::set-output name=valid::{valid_code}")
+        
+
```
Comment (vybhav72954) [inline .github/actions/PR_has_a_valid_issue/main.py:38]: Only 1 blank line at the end please

**Review by vybhav72954 — CHANGES_REQUESTED** [review]

**Comment by vybhav72954** [comment]
@XZANATOL I will do a one-final run from my side just to ensure everything is working. Will get this merge tomorrow.

I was just wondering, If the PR doesn't have the label, can you make it a draft PR as well, that functionality is a bit weird from GitHub's API, but pretty sure is possible. What do you think?

**Comment by XZANATOL** [comment]
@vybhav72954 hmmm, the action will either assign a `` PR:Ready-to-Review `` or `` PR:No-Issue `` label .. all what is needed is the carteria under which the PR will get drafted :thinking: `` gh `` CLI does support this change so It won't be a problem

**Comment by vybhav72954** [comment]
> @vybhav72954 hmmm, the action will either assign a `` PR:Ready-to-Review `` or `` PR:No-Issue `` label .. all what is needed is the carteria under which the PR will get drafted :thinking: `` gh `` CLI does support this change so It won't be a problem

Yeah so what i want is, that if the PR has no issue, or basically a `PR:No-Issue` label.

Just run more job to make it a draft.

**Comment by XZANATOL** [comment]
Easy one, but if the author changed the body and put a valid issue num.. shall the action revert the change?

**Comment by vybhav72954** [comment]
> Easy one, but if the author changed the body and put a valid issue num.. shall the action revert the change?

If that's possible, would be a great addition. If not, then modify the comment accordingly.

Something like this PR is drafted for now. After linking the issue properly make to ready for review.

Also as we have a completely separate label for properly edited PRs, i think marking it Ready won't be a problem?

**Comment by XZANATOL** [comment]
Gonna test it out and update you ASAP, and I think it won't be a problem to mark the PR ready as any edits that doesn't relate to the issue number will be from the mentor's side (and mentors can add `` PR:Under-Review `` label to indicate that it is under review process)

**Comment by XZANATOL** [comment]
@vybhav72954 I was looking for the right syntax of gh CLI command to mark a draft PR to ready and the opposite. What I've found is gh can convert a draft PR to ready-to-review but not the opposite. GitHub didn't support this yet - check [this](https://github.com/cli/cli/issues/740#issuecomment-630319524) -. I though about changing the metadata of the PR using `` gh api `` but it seems I can't edit the metadata using it. (also not supported)

**Comment by XZANATOL** [comment]
@vybhav72954 updates?

**Comment by vybhav72954** [comment]
> @vybhav72954 updates?

Yup as i thought. @HarshCasper This looks gud to me

**Review by vybhav72954 — APPROVED** [review]

**Comment by vybhav72954** [comment]
@XZANATOL Am pretty sure you can add an image as well right? Just upload it on an image hosting portal and paste the link with the job-creating comment.

**Comment by vybhav72954** [comment]
f
Can you make sure this image is a part of comment as well, and make sure to change the comment to something like, **It should look like this**

**Comment by XZANATOL** [comment]
I think yea, gonna prepare this rught away

**Comment by vybhav72954** [comment]
I will get this merge tonight for sure.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**